File: frontend/src/api/games.ts
import { apiFetch } from "./client";
import type { MatchingRound, MadLibsRound, ScramblerRound, SentenceCount, TuneInRound, AudioCardCount, ScrambleHarderRound } from "../types/game";

export function getMatchingRound(level: number): Promise<MatchingRound> {
  return apiFetch(`/api/games/matching?level=${level}`);
}

export function getMadLibsRound(level: number): Promise<MadLibsRound> {
  return apiFetch(`/api/games/madlibs?level=${level}`);
}

export function getScramblerRound(level: number): Promise<ScramblerRound> {
  return apiFetch(`/api/games/scrambler?level=${level}`);
}

export function getSentenceCount(level: number): Promise<SentenceCount> {
  return apiFetch(`/api/games/sentence-count?level=${level}`);
}

export function getTuneInRound(level: number): Promise<TuneInRound> {
  return apiFetch(`/api/games/tunein?level=${level}`);
}

export function getAudioCardCount(): Promise<AudioCardCount> {
  return apiFetch(`/api/games/audio-card-count`);
}

export function getScrambleHarderRound(level: number): Promise<ScrambleHarderRound> {
  return apiFetch(`/api/games/scramble-harder?level=${level}`);
}

File: backend/routers/games.py
from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.security import APIKeyHeader

from backend.models.game import MatchingRound, MadLibsRound, ScramblerRound, SentenceCount
from backend.services import game_service

_token_header = APIKeyHeader(name="x-trilingo-token", auto_error=False)

router = APIRouter(
    prefix="/api/games",
    tags=["games"],
    dependencies=[Depends(_token_header)],
)


@router.get("/matching", response_model=MatchingRound)
async def get_matching(level: int = Query(1, ge=1, le=3)):
    return await game_service.get_matching_round(level)


@router.get("/madlibs", response_model=MadLibsRound)
async def get_madlibs(level: int = Query(1, ge=1, le=3)):
    try:
        return await game_service.get_madlibs_round(level)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to generate round: {e}")


@router.get("/sentence-count", response_model=SentenceCount)
async def get_sentence_count(level: int = Query(1, ge=1, le=3)):
    return await game_service.get_sentence_count(level)


@router.get("/scrambler", response_model=ScramblerRound)
async def get_scrambler(level: int = Query(1, ge=1, le=3)):
    try:
        return await game_service.get_scrambler_round(level)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
